### PR TITLE
BAPL-1177/JBPM-6871 - Process Properties (part 2)

### PR DESCRIPTION
**Pull request by Josephblt**
https://issues.jboss.org/browse/BAPL-1177
https://issues.jboss.org/browse/JBPM-6871

Executable, Globals properties are supported now.
The imports property will no longer be a part of this PR. It can be tracked by this JIRA:
https://issues.jboss.org/browse/JBPM-8221

### CODE REVIEW

**Comment by Josephblt** [comment]
Jenkins execute full downstream build

**Comment by Josephblt** [comment]
Hey @romartin, @LuboTerifaj, if you guys could take a look, that be great. Thanks.

**Comment by Josephblt** [comment]
Jenkins execute full downstream build

File: kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-bpmn/kie-wb-common-stunner-bpmn-backend/src/test/java/org/kie/workbench/common/stunner/bpmn/backend/service/diagram/BPMNDirectDiagramMarshallerTest.java [hunk 1]
```diff
@@ -395,6 +405,40 @@ public void testUnmarshallProcessVariables() throws Exception {
...
+    @Test
+    @SuppressWarnings("unchecked")
+    public void testUnmarshallGlobalVariables() throws Exception {
+        Diagram<Graph, Metadata> diagram = unmarshall(BPMN_GLOBALVARIABLES);
+        assertDiagram(diagram,
+                      1);
+        assertEquals("Global Variables",
+                     diagram.getMetadata().getTitle());
+        GlobalVariables variables = null;
+        Iterator<Element> it = nodesIterator(diagram);
```
Comment (romartin) [inline kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-bpmn/kie-wb-common-stunner-bpmn-backend/src/test/java/org/kie/workbench/common/stunner/bpmn/backend/service/diagram/BPMNDirectDiagramMarshallerTest.java]: IMO better moving this logic out from the test scope, you can just create some helper method in some other or new class out there, and then just call it here. Also, IIRC I think I've seen this code in some other places, you know if that's being duplicated somewhere? If so, can you just refactor it to the call to the static helper method? Thx

File: kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-bpmn/kie-wb-common-stunner-bpmn-backend/src/main/java/org/kie/workbench/common/stunner/bpmn/backend/converters/customproperties/elements/GlobalVariablesElement.java [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+    }
+
+    @Override
+    public void setValue(BaseElement element, String value) {
+        setStringValue(element, value);
+    }
+
+    @Override
+    public void setStringValue(BaseElement element, String value) {
+        String[] variables = value.split(",");
```
Comment (romartin) [inline kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-bpmn/kie-wb-common-stunner-bpmn-backend/src/main/java/org/kie/workbench/common/stunner/bpmn/backend/converters/customproperties/elements/GlobalVariablesElement.java]: I would suggest to move to the use of Streams (stream().filter().map().forEach()) to make the code more readable

File: kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-bpmn/kie-wb-common-stunner-bpmn-backend/src/main/java/org/kie/workbench/common/stunner/bpmn/backend/converters/customproperties/elements/GlobalVariablesElement.java [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+            }
+        }
+    }
+
+    @Override
+    protected Optional<String> getStringValue(BaseElement element) {
+        List<ExtensionAttributeValue> extensionValues = element.getExtensionValues();
+        String globalVariables = null;
+
+        if (extensionValues != null && extensionValues.size() > 0) {
```
Comment (romartin) [inline kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-bpmn/kie-wb-common-stunner-bpmn-backend/src/main/java/org/kie/workbench/common/stunner/bpmn/backend/converters/customproperties/elements/GlobalVariablesElement.java]: I would suggest to move to the use of Streams (stream().filter().map().collect()) to make the code more readable. Use can for example use `Collectors.joining()` to generate the resulting String...

**Comment by LuboTerifaj** [comment]
Jenkins retest please.

**Comment by LuboTerifaj** [comment]
Jenkins execute full downstream build

**Comment by Josephblt** [comment]
Jenkins execute full downstream build

**Comment by Josephblt** [comment]
Hello @romartin. All set, did the code changes you asked. Thanks.

**Comment by romartin** [comment]
ping @LuboTerifaj plz - urgen

**Comment by LuboTerifaj** [comment]
Jenkins retest please.

**Comment by LuboTerifaj** [comment]
Jenkins execute full downstream build

**Comment by romartin** [comment]
good job and good Qs @LuboTerifaj , thanks :+1: 
@Josephblt your turn :)

**Comment by Josephblt** [comment]
Hello @LuboTerifaj. Yeah it is somewhat confusing the name of the PR. But I just copied the JIRA name, so... A little explanation. 

Target Namespace and Type Language properties were dropped as they were considered unnecessary by @krisv and @romartin.

Imports property is still necessary and is being developed separately due to time constraints. you can watch it here: [JBPM-8221 - Imports property for Process](https://issues.jboss.org/browse/JBPM-8221)

All of this is reported in the following related JIRA's:
[BAPL-1177 - Stunner - process properties (part 2)](https://issues.jboss.org/browse/BAPL-1177)
[JBPM-6871 - Designer - Process properties (part 2)](https://issues.jboss.org/browse/JBPM-6871)
[JBPM-8221 - Imports property for Process](https://issues.jboss.org/browse/JBPM-8221)

About the Global Variables:
In the legacy designer, global variables are not available in the I/O assignments editor.
The scope is global, available throughout the process
```
<bpmn2:process id="NewBPMN2diagram" ...>
     <bpmn2:extensionElements>
          <drools:global identifier="GV1" type="Boolean"/>
     </bpmn2:extensionElements>
</bpmn2:process>
```

Right now the Case Management editor supports Global Variables. I treated it as a "copy" of the Business Process editor.

**Comment by Josephblt** [comment]
Jenkins execute full downstream build

**Comment by Josephblt** [comment]
Jenkins execute full downstream build

**Comment by romartin** [comment]
ping @Josephblt @LuboTerifaj - **URGENT** to look at this pleaseeee

**Comment by Josephblt** [comment]
Jenkins execute full downstream build

**Comment by Josephblt** [comment]
Jenkins execute full downstream build

**Comment by Josephblt** [comment]
Heloo @romartin. Failing tests are not related to this PR. Ready to test.

**Comment by Josephblt** [comment]
Hey @LuboTerifaj. Yeah, it seems odd but that's the correct behaviour. 
If you take a look at this test:
https://github.com/wmedvede/jbpm/blob/08feab60acf139ce2f8af689d1e20878cceca070/jbpm-bpmn2/src/test/java/org/jbpm/bpmn2/IntermediateEventTest.java#L2144-L2144
you'll notice that is the exact behaviour that we have now. 

Thanks.

**Comment by romartin** [comment]
Hey @LuboTerifaj please let us know something, let's see if we can merge this for the release! Thanks!

**Comment by romartin** [comment]
ping @LuboTerifaj  @hasys  - please missing review here!!

**Comment by krisv** [comment]
@LuboTerifaj This seems to be correct behavior.  Note that it would not show up in the list of keys unless it is already set (and there are various ways of doing this).  I believe this seems to indicate it works as otherwise you would not be able to set the global if it's not known.  For example, if you define a java.util.List global called "list", can you in a script then add "list.add(...);" so that it compiles (it will probably result in a NPE unless you set the global first but that's fine)?

**Comment by romartin** [comment]
thanks @LuboTerifaj !!

**Pull request merged**